include/trieste/logging.h

```diff
@@ -0,0 +1,278 @@
+    // Defaults to stdout.
+    inline static std::function<void(std::stringstream&)> dump_callback{
+      [](std::stringstream& s) { std::cout << s.str() << std::flush; }};
+
+    // Delete copy constructor and assignment operator.
+    Log(const Log&) = delete;
+    Log& operator=(const Log&) = delete;
+
+    // Delete move constructor and assignment operator.
+    Log(Log&&) = delete;
+    Log& operator=(Log&&) = delete;
+
+    static bool active()
+    {
+      return Level <= detail::report_level;
+    }
+
+    SNMALLOC_FAST_PATH Log()
+    {
+      if (SNMALLOC_UNLIKELY(active()))
+        start();
+    }
+
+    SNMALLOC_FAST_PATH Log& operator<<(std::ostream& (*f)(std::ostream&)) &
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        operation(f);
+
+      return *this;
+    }
+
+    SNMALLOC_FAST_PATH Log& operator<<(std::ostream& (*f)(std::ostream&)) &&
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        operation(f);
+
+      return *this;
+    }
+
+    SNMALLOC_FAST_PATH Log& operator<<(detail::Indent) &
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        indent();
+      return *this;
+    }
+
+    SNMALLOC_FAST_PATH Log& operator<<(detail::Indent) &&
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        indent();
+      return *this;
+    }
+
+    SNMALLOC_FAST_PATH Log& operator<<(detail::Undent) &
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        undent();
+      return *this;
+    }
+
+    SNMALLOC_FAST_PATH Log& operator<<(detail::Undent) &&
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        undent();
+      return *this;
+    }
+
+    template<typename T>
+    SNMALLOC_FAST_PATH Log& operator<<(T&& t) &
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        append(std::forward<T>(t));
+      return *this;
+    }
+
+    template<typename T>
+    SNMALLOC_FAST_PATH Log& operator<<(T&& t) &&
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        append(std::forward<T>(t));
+      return *this;
+    }
+
+    SNMALLOC_FAST_PATH ~Log()
+    {
+      if (SNMALLOC_UNLIKELY(print))
+        end();
+    }
+  };
+
+  using Output = Log<detail::LogLevel::Output>;
+  using Error = Log<detail::LogLevel::Error>;
+  using Warn = Log<detail::LogLevel::Warn>;
+  using Info = Log<detail::LogLevel::Info>;
+  using Debug = Log<detail::LogLevel::Debug>;
+  using Trace = Log<detail::LogLevel::Trace>;
+
+#ifdef TRIESTE_EXPOSE_LOG_MACRO
+// This macro is used to expose the logging to uses in a way that
+// guarantees no evaluation of the pipe sequence:
+//   LOG(Info) << "Hello " << "World" << fib(23);
+// would not evaluate fib(23) if Info is not enabled.
+// Where as the pure C++ version
+//   logging::Info() << "Hello " << "World" << fib(23);
+// would be required to evaluate fib(23) even if Info is not enabled.
+#  define LOG(param) \
+    if (SNMALLOC_UNLIKELY( \
+          trieste::logging::Log< \
+            trieste::logging::detail::LogLevel::param>::active())) \
+    trieste::logging::Log<trieste::logging::detail::LogLevel::param>()
```

Comment:
Just to be safe, may want to add braces around this

Reply:
Where do you mean?  If it is `{` and `}` around the whole lot, then that will do the wrong thing.  The macro is intentionally leaving the scope of the `if` open:
```C++
LOG(Info) << fib(23)
  << fib(21);
```
should rewrite to
```C++
if (...)
trieste::logging::Log<trieste::logging::detail::LogLevel::Info>() << fib(23)
  << fib(21);
```
which is really
```C++
if (...)
{
  trieste::logging::Log<trieste::logging::detail::LogLevel::Info>() << fib(23)
    << fib(21);
}
```
This is obviously horrible, and the pure C++ one works well in most cases.

Reply:
I misread the macro, I thought it encased the entire production but rather it is just the head, so you need to do it the way you do here.